aea/runner.py

```diff
@@ -0,0 +1,234 @@
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains the implementation of AEA multiple instances runner."""
+import asyncio
+from abc import ABC, abstractmethod
+from asyncio.events import AbstractEventLoop
+from asyncio.tasks import FIRST_EXCEPTION
+from concurrent.futures.thread import ThreadPoolExecutor
+from threading import Thread
+from typing import Awaitable, Dict, Optional, Sequence, Type
+
+from aea.aea import AEA
+
+
+class BaseAEAExecutor(ABC):
+    """Base AEA executor to impmlement aea executors to run multiple agents."""
+
+    def __init__(self, agents: Sequence[AEA]) -> None:
+        """
+        Init executor.
+
+        :param agents: sequence of AEA instances to run.
+        """
+        self._agents: Sequence[AEA] = agents
+        self._is_running: bool = False
+        self._agent_task: Dict[AEA, Awaitable] = {}
+        self._task_agent: Dict[Awaitable, AEA] = {}
+        self._loop: AbstractEventLoop = asyncio.new_event_loop()
+        self._set_executor_pool()
+
+    @property
+    def is_running(self) -> bool:
+        """Return running state of the executor."""
+        return self._is_running
+
+    def start(self) -> None:
+        """Start agents."""
+        self._is_running = True
+        self._start_agents()
+        self._loop.run_until_complete(self._wait_tasks())
+
+    def stop(self) -> None:
+        """Stop agents."""
+        self._is_running = False
+        for agent in self._agents:
+            self._stop_agent(agent)
+        if not self._loop.is_running():
+            self._loop.run_until_complete(self._wait_tasks(skip_exceptions=True))
+
+    def _start_agents(self) -> None:
+        """Start agents tasks."""
+        for agent in self._agents:
+            task = self._start_agent(agent)
+            self._agent_task[agent] = task
+            self._task_agent[task] = agent
+
+    async def _wait_tasks(self, skip_exceptions: bool = False) -> None:
+        """
+        Wait agents tasks to complete.
+
+        :param skip_exceptions: skip exceptions if raised in tasks
+        """
+        done, pending = await asyncio.wait(
+            self._task_agent.keys(), return_when=FIRST_EXCEPTION
+        )
+
+        async def wait_task(task):
+            try:
+                await task
+            except Exception:
+                if not skip_exceptions:
+                    raise
+
+        for task in done:
+            await wait_task(task)
+
+        if pending:
+            done, _ = await asyncio.wait(pending)
+            for task in done:
+                await wait_task(task)
```

Comment:
Should `_wait_tasks` not be moved into the concrete class?

Reply:
Ah I see, so we use asyncio to run the executor pool or the agents directly..fine